File: 706_design_hashmap.java
class MyHashMap {
    Integer[] bucket;
    final int SIZE = 1000000;
    
    /** Initialize your data structure here. */
    public MyHashMap() {
        bucket = new Integer [SIZE];
    }
    
    /** value will always be non-negative. */
    public void put(int key, int value) {
        bucket[key] = new Integer(value);
    }
    
    /** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
    public int get(int key) {
        if (bucket[key] == null) {
            return -1;   
        }
        return bucket[key];
    }
    
    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    public void remove(int key) {
        if (bucket[key] != null) {
            bucket[key] = null;
        }
    }
}

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap obj = new MyHashMap();
 * obj.put(key,value);
 * int param_2 = obj.get(key);
 * obj.remove(key);
 */

File: python/706_design_hashmap.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None


class MyHashMap:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.capacity = 1000
        self.bucket = [None for _ in range(self.capacity)]

    def put(self, key: int, value: int) -> None:
        """
        value will always be non-negative.
        """
        index = key % self.capacity
        if not self.bucket[index]:
            self.bucket[index] = Node(key, value)
        else:
            # collesion, iterate the linked list and find the value
            curr = self.bucket[index]
            while curr:
                if curr.key == key:
                    curr.value = value
                    return
                if not curr.next: break
                curr = curr.next
            curr.next = Node(key, value)

    def get(self, key: int) -> int:
        """
        Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key
        """
        index = key % self.capacity
        curr = self.bucket[index]
        while curr:
            if curr.key == key:
                return curr.value
            curr = curr.next
        return -1

    def remove(self, key: int) -> None:
        """
        Removes the mapping of the specified value key if this map contains a mapping for the key
        """
        index = key % self.capacity
        curr = prev = self.bucket[index]
        if not curr: return
        if curr.key == key:
            self.bucket[index] = curr.next
        else:
            curr = curr.next
            while curr:
                if curr.key == key:
                    prev.next = curr.next
                    break
                else:
                    curr, prev = curr.next, prev.next



# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)
